### PR TITLE
Transactions and holding result-sets

### PR DESCRIPTION
**Pull request by bastbijl**
- Add getters and setters for WeaverQuery to keep the result set open
- Add route to close a connection that was kept open
- Extra tests for querying with an open connection

I certify that:
- [x] The automated build passes on the branch this pull request is from
- [x] My change is documented in the Changelist.md
- [ ] Any development process additions/changes are documented in the README.md
- [ ] Any sdk usage additions/changes are documented [here](https://github.com/weaverplatform/weaver-docs/blob/master/pages/developers/reference/weaver-sdk-js.md "Weaver-docs")
- [x] Any new dockerimages this build depends on are properly tagged on github
- [x] I shall not merge until the travis build is green and the PR is accepted by
  another developer
- [x] I shall promptly fix/reply to comments, as I know PRs are not :fire: and
  forget
- [x] I shall keep my branch updated with any intermediate changes to the
  target branch
- [x] This is a quality PR
- [x] I got to the end of this list

### CODE REVIEW

File: src/WeaverQuery.coffee [hunk 1]
```diff
@@ -330,6 +334,14 @@ class WeaverQuery
     @_limit = limit
     @
 
+  keepOpen: (keepOpen=true) ->
+    @_keepOpen = keepOpen
+    @
+
+  setCursorName: (name) ->
+    @_cursor = name if name?
```
Comment (gvdent) [inline src/WeaverQuery.coffee]: What is the reasoning behind requiring the user to define this as opposed to using a generated cuid?

File: test/WeaverQuery.test.coffee [hunk 2]
```diff
@@ -1605,3 +1605,72 @@ describe 'WeaverQuery Test', ->
...
+  describe 'query nodes in one stream', ->
+
+    before ->
+      wipeCurrentProject()
+      .then(->
+        Weaver.Node.batchSave((new Weaver.Node(i) for i in [0...5]))
+      )
+
+    it 'should find 5 nodes in steps of 1', ->
+
```
Comment (gvdent) [inline test/WeaverQuery.test.coffee]: Please don't use superfluous blank lines. Don't start the body of a function with a blank line, within function bodies a single blank line can be used to improve readability. (https://github.com/sysunite/coffeescript-style-guide/#blank-lines)

File: test/WeaverQuery.test.coffee [hunk 3]
```diff
@@ -1605,3 +1605,72 @@ describe 'WeaverQuery Test', ->
...
+      query = new Weaver.Query()
+      .keepOpen()
+      .setCursorName('abc')
+      .limit(1)
+
+      query.find()
```
Comment (gvdent) [inline test/WeaverQuery.test.coffee]: Since tests illustrate proper usage of functionality, please use the proper construct to close the cursor as well.

I'd suggest enforcing _always_ placing these in a `finally` block so they are closed even when an error occurs.

File: test/WeaverQuery.test.coffee (hunk 2 above)
Comment (bastbijl) [reply test/WeaverQuery.test.coffee]: done

File: Changelist.md [hunk 4]
```diff
@@ -5,6 +5,8 @@
 - Adding test to validate id names on cloning projects
 - Added created by constraint to weaver query
 - Fix checking value for date. Don't throw error on js Date objects
+- Add getters and setters for WeaverQuery to keep the result set open
+- Add route to close a connection that was kept open
```
Comment (gvdent) [inline Changelist.md]: Needs:
- own version number
- own changelist header
- line addresssing transaction support being added.

File: package.json [hunk 5]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "weaver-sdk",
-  "version": "11.2.0",
+  "version": "11.2.0-alpha.0-transactions",
```
Comment (gvdent) [inline package.json]: New feature, so 11.3.0, version's dont go backward (this would introduce an alpha of an already released version).

File: src/Weaver.coffee [hunk 6]
```diff
@@ -66,6 +66,15 @@ class Weaver
   useProject: (project) ->
     @coreManager.currentProject = project
 
+  startTransaction: ->
+    if @coreManager.currentTransaction?
+      throw new Error("There is already an active transaction, first commit or rollback that one")
```
Comment (gvdent) [inline src/Weaver.coffee]: Dunglish, "A transaction is active already.." or something.

File: test-server/docker-compose.yml [hunk 7]
```diff
@@ -6,7 +6,7 @@ services:
     - "5432"
 
   postgresql-connector:
-    image: sysunite/weaver-database-postgresql:4.14.0
+    image: sysunite/weaver-database-postgresql:4.14.0-alpha.5-transactions
```
Comment (gvdent) [inline test-server/docker-compose.yml]: Reverting to an alpha from a stable looks weird, make sure the versions align correctly.

File: test/WeaverNode.test.coffee [hunk 8]
```diff
@@ -1277,6 +1277,99 @@ describe 'WeaverNode test', ->
...
+    ).then((nodes)->
+      expect(nodes.length).to.equal(1)
+      transaction.rollback()
+    ).then(->
+      expect(Weaver.getCoreManager().currentTransaction).to.be.undefined
+      new Weaver.Query().restrict(node).find()
+    ).then((nodes)->
+      expect(nodes.length).to.equal(0)
+    ).catch(->
+      transaction.rollback()
```
Comment (gvdent) [inline test/WeaverNode.test.coffee]: I don't this the catch is needed here.

If this is not required, either fail an assert here, or don't catch it at all and let mocha handle the failure.

File: test/WeaverNode.test.coffee [hunk 9]
```diff
@@ -1277,6 +1277,99 @@ describe 'WeaverNode test', ->
...
+    ).then((nodes)->
+      expect(nodes.length).to.equal(1)
+      transaction.commit()
+    ).then(->
+      expect(Weaver.getCoreManager().currentTransaction).to.be.undefined
+      new Weaver.Query().restrict(node).find()
+    ).then((nodes)->
+      expect(nodes.length).to.equal(1)
+    ).catch(->
+      transaction.rollback()
```
Comment (gvdent) [inline test/WeaverNode.test.coffee]: Same comment as previous catch. Adding catches like this make tests that would otherwise fail pass and swallow the error.

File: test/WeaverNode.test.coffee [hunk 10]
```diff
@@ -1277,6 +1277,99 @@ describe 'WeaverNode test', ->
...
+    .then(->
+      Weaver.getInstance().startTransaction()
+    ).then((trx)->
+      transaction = trx
+      node.set('age', 43)
+      node.save()
+    ).then(->
+      Weaver.getCoreManager().currentTransaction = undefined
+      node.set('age', 33)
+      node.save().should.be.rejectedWith('"replaced" violates not-null constraint')
```
Comment (gvdent) [inline test/WeaverNode.test.coffee]: This is going to be such an obscure error :) Nice to have a test for it! :100:

File: test/WeaverQuery.test.coffee [hunk 11]
```diff
@@ -1619,3 +1603,83 @@ describe 'WeaverQuery Test', ->
...
+      )
+
+    it 'should find 5 nodes in steps of 2', ->
+      query = null
+      transaction = null
+      Weaver.getInstance().startTransaction()
+      .then((trx)->
+        transaction = trx
+        query = new Weaver.Query()
+        .keepOpen()
```
Comment (gvdent) [inline test/WeaverQuery.test.coffee]: Does the added startTransaction() to all these tests now mean that keepOpen is only supported within a transaction?

If so, a test of the behaviour without a startTransaction() should be added.

imho if `keepOpen` is used, it should start a transaction automatically if none is set so we're not forced to do the `startTransaction().then(->` spiel every time we want to keep a result set open.

File: Changelist.md (hunk 4 above)
Comment (bastbijl) [reply Changelist.md]: done

File: package.json (hunk 5 above)
Comment (bastbijl) [reply package.json]: done

File: src/Weaver.coffee (hunk 6 above)
Comment (bastbijl) [reply src/Weaver.coffee]: done

File: test-server/docker-compose.yml (hunk 7 above)
Comment (bastbijl) [reply test-server/docker-compose.yml]: done

File: test/WeaverNode.test.coffee (hunk 9 above)
Comment (bastbijl) [reply test/WeaverNode.test.coffee]: done

File: test/WeaverNode.test.coffee (hunk 8 above)
Comment (bastbijl) [reply test/WeaverNode.test.coffee]: done

File: test/WeaverNode.test.coffee (hunk 10 above)
Comment (bastbijl) [reply test/WeaverNode.test.coffee]: 🦹‍♂️

File: test/WeaverQuery.test.coffee (hunk 11 above)
Comment (bastbijl) [reply test/WeaverQuery.test.coffee]: Yes, and I agree. Let me add that test and also add the implicit transaction starting.

Comment (bastbijl) [reply test/WeaverQuery.test.coffee]: done

File: test/WeaverQuerySingleNetwork.test.coffee [hunk 12]
```diff
@@ -328,23 +328,17 @@ describe 'WeaverQuery with single Network', ->
...
 
   it 'should have count and res.length be equal', ->
-    q = new Weaver.Query()
     Promise.all([
-      q.find().then((res) ->
-        res.length
-      ),
-      q.count()
+      new Weaver.Query().find().then((res) -> res.length),
+      new Weaver.Query().count()
```
Comment (gvdent) [inline test/WeaverQuerySingleNetwork.test.coffee]: This change scares me a little... Can query objects no longer be reused now?

File: package.json [hunk 13]
```diff
@@ -1,15 +1,15 @@
...
   "author": {
     "name": "Mohamad Alamili",
     "url": "https://github.com/weaverplatform/weaver-sdk-js",
     "email": "mohamad@sysunite.com"
   },
   "com_weaverplatform": {
-    "requiredConnectorVersion": "^4.14.0",
-    "requiredServerVersion": "^3.13.0"
+    "requiredConnectorVersion": "4.15.0-beta.0",
+    "requiredServerVersion": "3.17.0-beta.0"
```
Comment (gvdent) [inline package.json]: You can set the required versions as the stable targets, so `^4.15.0` and `^3.17.0`, that way you do not need to change it later.

Version checks are disabled for prerelease versions anyway.

File: test/WeaverQuerySingleNetwork.test.coffee (hunk 12 above)
Comment (bastbijl) [reply test/WeaverQuerySingleNetwork.test.coffee]: fixed

File: package.json (hunk 13 above)
Comment (bastbijl) [reply package.json]: done

File: src/CoreManager.coffee [hunk 14]
```diff
@@ -263,7 +266,7 @@ class CoreManager
 
   query: (query) ->
     target = query.target
-    query  = _.omit(query, ['model', 'context', 'target', 'preferredConstructor', 'constructorFunction'])
+    delete query.target
```
Comment (gvdent) [inline src/CoreManager.coffee]: Please revert this change. 

I understand the reason this may not lead to issues straight away is that query now does the preserialize clone, but keeping the code the same keeps the code more fault tolerant. Now the coreManager changes the object passed into it, which is an bad pattern in oop.

File: test/WeaverGraph.test.coffee [hunk 15]
```diff
@@ -20,7 +20,7 @@ describe 'WeaverGraph support', ->
       Weaver.Node.load(c.id()).should.be.rejected
 
     it 'should not load a node if the wrong graph is specified', ->
-      Weaver.Node.load(b.id(), aInGraph.getGraph()).should.be.rejected
+      Weaver.Node.loadFromGraph(b.id(), aInGraph.getGraph()).should.be.rejected
```
Comment (gvdent) [inline test/WeaverGraph.test.coffee]: Thanks :+1:
